### DISCUSSION TITLE
Can form be embedded in a vertical viewport?

### DISCUSSION THREAD

**Discussion by amano-kenji** [jtdaugherty/brick#455] [Q&A] [answered]
What if form is too tall to fit in its given height? Does a form element make a visibility request?

Maybe, I'm supposed to split one form into multiple forms and make a visibility request for each one?

**Comment by jtdaugherty** [comment]
> What if form is too tall to fit in its given height?

This is a bit unrelated to the viewport question, but for this question, the answer is that the form will only render what it can, from to bottom.

> Does a form element make a visibility request?

No, but it would be pretty easy to add that in your application by using [formFocus](https://hackage.haskell.org/package/brick-1.6/docs/Brick-Forms.html#v:formFocus) and [visible](https://hackage.haskell.org/package/brick-1.6/docs/Brick-Widgets-Core.html#v:visible) to focus the currently-focused field in your drawing routine.

**Reply by amano-kenji** [reply]
`visible` requires `Widget n`. I don't see a way to extract `Widget n` from `FocusRing n`.

**Reply by jtdaugherty** [reply]
What I mean is that you'd augment the `renderForm` call with a call to `visible`, then wrap the whole thing in `viewport`. That will cause the viewport to scroll (if necessary) to make the selected form field visible. To get the required resource name, you'd use `formFocus` and [focusGetCurrent](https://hackage.haskell.org/package/brick-1.6/docs/Brick-Focus.html#v:focusGetCurrent) to get the name of the focused field and pass that to `visible`. So something like this:

```haskell
draw :: s -> Widget n
draw s =
    viewport ... $
    makeVisible $
    renderForm ...
    where
        makeVisible = case focusGetCurrent (formFocus f) of
            Nothing -> id
            Just n -> visible n 
```

**Reply by amano-kenji** [reply]
`visible :: Widget n -> Widget n`. `n` is not `Widget n`. Thus, `visible n` is wrong.

**Reply by amano-kenji** [reply]
`makeVisible :: Ord n => n -> EventM n s ()`, so it cannot be used in rendering functions. But, you can use this in event handlers.

**Reply by amano-kenji** [reply]
I think I may have to render each form field separately. I don't know how to do that, yet.

**Reply by amano-kenji** [reply]
> FormField
formFieldRender :: Bool -> b -> Widget n
A function to render this form field. Parameters are whether the field is currently focused, followed by the field state.

If I can customize this, then I can use the boolean parameter to decide whether to wrap a widget in `visible`.

**Comment by jtdaugherty** [comment, marked as answer]
> visible :: Widget n -> Widget n. n is not Widget n. Thus, visible n is wrong.

Ah yes, sorry about that. `visible w` means "make `w` visible if you can."

> makeVisible :: Ord n => n -> EventM n s (), so it cannot be used in rendering functions. But, you can use this in event handlers.

I think this is actually what I was thinking of, and I forgot the exact type. Since you can just as easily access the form state in your event handler, this is probably the best way to go.

**Reply by amano-kenji** [reply]
Can I customize `formFieldRender :: Bool -> b -> Widget n`?

**Reply by jtdaugherty** [reply]
You could do that when constructing arguments to [the various lower-level field constructors](https://hackage.haskell.org/package/brick-1.6/docs/Brick-Forms.html#g:4).

**Reply by amano-kenji** [reply]
https://hackage.haskell.org/package/brick-1.6/docs/Brick-Forms.html#v:editField and other functions don't give me `Bool`.

If there are lenses for FormFieldState and FormField, then I can update `formFieldRender` with something like

```
addVisibilityRequest :: FormFieldState s e n -> FormFieldState s e n
addVisibilityRequest fs =
  fs & formFields . traverse . formFieldRender %~ \r -> \f s -> if f then visible (r f s) else r f s
```

Maybe, you can add this function to `Brick.Forms`.

**Reply by jtdaugherty** [reply]
> If there are lenses for FormFieldState and FormField

There are no lenses in the Form API. I'd rather not add a function specific to visibility requests if there is another way to do it (which I described above).

**Reply by amano-kenji** [reply]
Requesting visibility in EventM is going to require boilerplate in multiple places. Doing it in RenderM is a lot cleaner.

For example, `visible` should be called in resize event and keyboard event.

**Reply by jtdaugherty** [reply]
> For example, visible should be called in resize event and keyboard event.

Yeah, that should be just two places.

In any case, I'll update the form implementation to issue a visibility request for the focused field and include that in the next release.

**Reply by jtdaugherty** [reply]
In case you want to use the change now instead of waiting for a release, it's on `master` in cd8fce14759cfc629732d61a2f13a5a1a4e72e8c.

**Reply by amano-kenji** [reply]
Much appreciated.

**Reply by amano-kenji** [reply]
Ater research, I discovered that lens cannot penetrate existential data types like `FormFieldState`.

I should limit existential data types to terminal data types that have no product data types.

DerivieGeneric for existential data types implies generic lens for existential data types. Thus, I consider it an open-ended research problem.
